### PR TITLE
Option to use arduino-cli instead of Arduino IDE

### PR DESCRIPTION
**Pull request by giuliof**
Solves the issue #856 
* Adds a checkbox to the configuration interface to choose between standard Arduino and the CLI version ;
* Is necessary to explicitly set "arduino-cli" in command path;
* the support for Windows and MacOS is not complete.

### CODE REVIEW

**Comment by NikilRagav** [comment]
Wait so is this not merged into the extension?

**Comment by giuliof** [comment]
> Wait so is this not merged into the extension?

Does not seem so. Furthermore I developed and tested Arduino-CLI integration only on Linux platform. A bit more work is needed for Windows and OS X.

**Comment by sbhklr** [comment]
@giuliof Thanks for your work! Can you point out what exactly needs to be done for MacOS and windows? I'd be happy to help on the MacOS side.

**Comment by giuliof** [comment]
@sbhklr If I remember well, the function `validateArduinoPath` must be readjusted in `darwin.ts` and `win32.ts`, respectively for MacOS and Windows.
This function searches for the effective presence of Arduino executable within the system, throwing an error otherwise:
* on Linux side I already handled the switch between arduino and arduino-cli executables (see [`linux.ts`](https://github.com/giuliof/vscode-arduino/blob/b70a95ef982ac05a38e706cdf58e10b13b6661ba/src/common/sys/linux.ts#L24))
* on MacOs it searches for `Arduino.app` inside the Applications folder. A switch must be added to make it search for the appropriate Arduino-CLI app file;
* on Windows it searches inside registry keys and I have not even tried to understand how to add the support 😅

**Comment by MarcLV** [comment]
Please review this! :) I would love this feature

**Review by MarcLV — APPROVED** [review]

**Comment by MarcLV** [comment]
This feature update is basically what this package is suppose to be used for in my opinion ^^

**Comment by sbhklr** [comment]
@MarcLV It needs a bit more polishing. As indicated above Windows and MacOS support yet needs to be completed. I've started to work on it.

**Comment by MarcLV** [comment]
@sbhklr Aha, that make sense! Thank you very much for looking into that :)

File: src/common/sys/win32.ts
Comment (MarcLV) [inline src/common/sys/win32.ts:31]:
```suggestion
    return fileExistsSync(path.join(arduinoPath, isArduinoCli ? "arduino-cli.exe" : "arduino_debug.exe"));

```

File: src/common/sys/darwin.ts
Comment (MarcLV) [inline src/common/sys/darwin.ts:22]:
```suggestion
    return fileExistsSync(path.join(resolveMacArduinoAppPath(arduinoPath), isArduinoCli ? "arduino-cli" : "/Contents/MacOS/Arduino"));

```

**Review by MarcLV — CHANGES_REQUESTED** [review]
I think this will solve it for PC/Mac

File: src/common/sys/darwin.ts
Comment (sbhklr) [reply src/common/sys/darwin.ts:22]: Looks promising, I'll test it 👍

Comment (MarcLV) [reply src/common/sys/darwin.ts:22]: I appreciate it, man! :fist:

Comment (MarcLV) [reply src/common/sys/darwin.ts:22]: You will have to change your arduino.path to where the arduino-cli is installed tho

**Comment by MarcLV** [comment]
Did you have time to check it out? 🥰

**Comment by MarcLV** [comment]
Hmmm, I dont understand the logs, can someone help me?

**Comment by giuliof** [comment]
> Did you have time to check it out? smiling_face_with_three_hearts

Sorry for the delay, I was quite busy these weeks.
I launched the test on my local machine and seems to run without errors. Idk why CI fails both on Linux and Windows, any suggestions?

**Comment by MarcLV** [comment]
> > Did you have time to check it out? smiling_face_with_three_hearts
> 
> Sorry for the delay, I was quite busy these weeks.
> I launched the test on my local machine and seems to run without errors. Idk why CI fails both on Linux and Windows, any suggestion?

Can you try and rerun the build? The logs are not showing that much..

**Comment by MarcLV** [comment]
Cheers! Thanks again @giuliof for adding this!
Now we just need a review for this 👍

**Comment by MarcLV** [comment]
Hey @sbhklr ! can you please have a look at this PR? :)

**Comment by BuddhiAbeyratne** [comment]
This PR literally solves around half a dozen issues pointing at the same error in diffrent ways. it would be greatly benificial if we can get this approved and merged as soon as possible helping vscode become a solid workflow choise for arduino development even further. 

microsoft/vscode#106500
#1074 
#1097 
#922 

are the latest issues i found laying around but this is steming from node js use and there are so many others that are old and solved with package updates 
the fact that people tout downgrading vscode to solve the problem is also bad altogether. 
It would be really helpfull if we can get this reviewed by the core team and merged.

File: src/arduino/arduino.ts
Comment (maddogjt) [inline src/arduino/arduino.ts:143]: Not including the sketch path here breaks breaks compilation of sketches in subfolders of the workspace.

File: src/arduino/arduino.ts
Comment (maddogjt) [inline src/arduino/arduino.ts:293]:
```suggestion
                args.push("--build-path", outputPath);

```

File: src/arduino/arduino.ts
Comment (maddogjt) [inline src/arduino/arduino.ts:217]: Upload for the cli should probably use 'arduino-cli compile --upload ...'

File: src/arduino/arduino.ts
Comment (maddogjt) [inline src/arduino/arduino.ts:144]: To maintain matching behavior with IDE version, this should probably use the compile comand with --upload flag, ensuring that code is both compiled and uploaded

**Review by maddogjt — CHANGES_REQUESTED** [review]
I was unable to verify or upload using the current changes.  I've sent some PR's with potential patches for the issues I encountered.

**Comment by robotdad** [comment]
I have PR in progress to update the current serial issue. Once I get that complete into the develop branch I'll update this one from there and work on getting a proper review started.

**Comment by NikilRagav** [comment]
Did this go through?

I'd love to start having my students use VS Code again instead of Arduino
IDE so that they can use code complete etc, but r n it's too buggy (random
unexpected red squigglies, the extension server crashes, etc) without these
changes.

**Comment by MarcLV** [comment]
Bump

**Review by adiazulay — COMMENTED** [review]
This looks really close, I've been working through the main features to make sure they work.

- [x] Verify (compile)
- [x] Upload (compile --upload)
(only undesired behavior here is the default build path is the same folder as the sketch, v0.14 of the cli has a clean build option that should take care of this.)
- [ ] Upload using programmer
(this seems to mostly work, but I haven't been able to test it e2e)
- [ ] Install Boards (cores)
(cores install correctly, but when I rebuild the extension some of the disappear. I'm investing this now)
- [x] Install Libraries

I'm hoping to be able to merge this into develop today, and get an rc out.

File: src/arduino/arduino.ts
Comment (sebromero) [reply src/arduino/arduino.ts]: The path of the sketch can be passed as a parameter but as you said the containing folder needs have the same name as the .ino file. The following two commands yield the same result
```
arduino-cli compile Disco/Disco.ino -b arduino:avr:uno
arduino-cli compile Disco -b arduino:avr:uno
cd Disco; arduino-cli compile -b arduino:avr:uno
```

**Comment by maddogjt** [comment]
@adiazulay I had previously sent a few PRs to the giuliof:testing/arduino-cli branch to address some of the issues mentioned I had brought up.

File: src/arduino/programmerManager.ts
Comment (sebromero) [inline src/arduino/programmerManager.ts:62]: Instead of duplicating all the switch cases, how about:
```
let prefix = "";
if(!this._settings.useArduinoCli) prefix = "arduino:";
```
And then for the programmer value just append it e.g.
`this._programmervalue = prefix + "avrisp"`

**Comment by adiazulay** [comment]
I'm going to wait until next week to merge this, there's good discussion and a bit more work left to do.

File: src/arduino/arduino.ts
Comment (adiazulay) [reply src/arduino/arduino.ts]: I added a new compile arg to upload(), set to true by default. This way we keep the functionality the same as the CLI, but also provide an option to upload without compiling

File: src/arduino/arduino.ts
Comment (adiazulay) [reply src/arduino/arduino.ts]: I refactored this method, upload using programmer is now an option within the upload method.

**Review by adiazulay — APPROVED** [review]

**Comment by BuddhiAbeyratne** [comment]
So glad this is finally going in 👍

**Comment by sebromero** [comment]
@adiazulay Thank you for your work on this! Do we have an option in the UI to just upload the lastest binary or is it only in the API for now?

**Comment by ubidefeo** [comment]
@NikilRagav 
>I'd love to start having my students use VS Code again instead of Arduino
IDE so that they can use code complete etc, but r n it's too buggy (random
unexpected red squigglies, the extension server crashes, etc) without these
changes.

I don't think this implementation will offer any additional advantage for code-intelligence.
On the other hand we're about to release an updated version of the new Arduino IDE with a pretty solid Language Server and code-assist.
It should be merged before the holidays :)

**Comment by adiazulay** [comment]
@sebromero uploading the latest binary is not exposed in the UI yet, but it's available in the command palette as ```Arduino CLI: Upload```

**Pull request merged**
